plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Bitwise.hs

```diff
@@ -0,0 +1,564 @@
+shiftNegClearHigh = property $ do
+  bs <- forAllByteString1
+  let bitLen = 8 * BS.length bs
+  n <- forAll . Gen.integral . Range.linear 1 $ bitLen - 1
+  i <- forAll . Gen.integral . Range.linear 0 $ n - 1
+  let lhsInner = mkIterAppNoAnn (builtin () PLC.BitwiseShift) [
+        mkConstant @ByteString () bs,
+        mkConstant @Integer () (fromIntegral . negate $ n)
+        ]
+  let lhs = mkIterAppNoAnn (builtin () PLC.ReadBit) [
+        lhsInner,
+        mkConstant @Integer () (fromIntegral $ bitLen - i - 1)
+        ]
+  evaluateAndVerify (mkConstant @Bool () False) lhs
+
+rotateRollover :: Property
+rotateRollover = property $ do
+  bs <- forAllByteString
+  let bitLen = 8 * BS.length bs
+  i <- forAll . Gen.integral . Range.linear (negate 512) $ 512
+  let lhs = mkIterAppNoAnn (builtin () PLC.BitwiseRotate) [
+        mkConstant @ByteString () bs,
+        mkConstant @Integer () (case signum i of
+                                  (-1) -> (negate . fromIntegral $ bitLen) + i
+                                  _    -> fromIntegral bitLen + i)
+        ]
+  let rhs = mkIterAppNoAnn (builtin () PLC.BitwiseRotate) [
+        mkConstant @ByteString () bs,
+        mkConstant @Integer () i
+        ]
+  let compareExp = mkIterAppNoAnn (builtin () PLC.EqualsByteString) [
+        lhs,
+        rhs
+        ]
+  evaluateAndVerify (mkConstant @Bool () True) compareExp
+
+rotateMoveBits :: Property
+rotateMoveBits = property $ do
+  bs <- forAllByteString1
+  let bitLen = 8 * BS.length bs
+  i <- forAll . Gen.integral . Range.linear 0 $ bitLen - 1
+  j <- forAll . Gen.integral . Range.linear (negate 256) $ 256
+  let lhs = mkIterAppNoAnn (builtin () PLC.ReadBit) [
+        mkConstant @ByteString () bs,
+        mkConstant @Integer () (fromIntegral i)
+        ]
+  let rhsRotation = mkIterAppNoAnn (builtin () PLC.BitwiseRotate) [
+        mkConstant @ByteString () bs,
+        mkConstant @Integer () (fromIntegral j)
+        ]
+  let rhsIndex = mkIterAppNoAnn (builtin () PLC.ModInteger) [
+        mkConstant @Integer () (fromIntegral $ i + j),
+        mkConstant @Integer () (fromIntegral bitLen)
+        ]
+  let rhs = mkIterAppNoAnn (builtin () PLC.ReadBit) [
+        rhsRotation,
+        rhsIndex
+        ]
+  evaluateTheSame lhs rhs
+
+csbRotate :: Property
+csbRotate = property $ do
+  bs <- forAllByteString
+  i <- forAll . Gen.integral . Range.linear (negate 512) $ 512
+  let lhs = mkIterAppNoAnn (builtin () PLC.CountSetBits) [
+        mkConstant @ByteString () bs
+        ]
+  let rhsInner = mkIterAppNoAnn (builtin () PLC.BitwiseRotate) [
+        mkConstant @ByteString () bs,
+        mkConstant @Integer () i
+        ]
+  let rhs = mkIterAppNoAnn (builtin () PLC.CountSetBits) [
+        rhsInner
+        ]
+  let compareExp = mkIterAppNoAnn (builtin () PLC.EqualsInteger) [
+        lhs,
+        rhs
+        ]
+  evaluateAndVerify (mkConstant @Bool () True) compareExp
+
+-- Helpers
+
+forAllByteString :: PropertyT IO ByteString
```

Comment:
I think these generators are duplicated in Laws.hs.  Is it worth just having one copy of each? They're not very big, so I don't really mind.